firefox-ios/Client/Frontend/Browser/WebView/PullRefreshView.swift

```diff
@@ -0,0 +1,323 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+import Common
+
+class PullRefreshView: UIView,
+                       ThemeApplicable {
+    private struct UX {
+        static let blinkProgressViewStandardThreshold: CGFloat = 80.0
+        static let progressViewPadding: CGFloat = 28.0
+        static let progressViewSize: CGFloat = 40.0
+        static let progressViewAnimatedBackgroundSize: CGFloat = 30.0
+        static let progressViewAnimatedBackgroundBlinkTransform = CGAffineTransform(scaleX: 2.1, y: 2.1)
+        static let progressViewAnimatedBackgroundFinalAnimationTransform = CGAffineTransform(scaleX: 15.0, y: 15.0)
+        static let easterEggSize = CGSize(width: 80.0, height: 100.0)
+        static let easterEggDelayInSeconds: CGFloat = 4.0
+    }
+
+    private let onRefreshCallback: VoidReturnCallback
+    private lazy var progressView: UIImageView = .build { view in
+        view.image = UIImage(named: StandardImageIdentifiers.Large.arrowClockwise)?.withRenderingMode(.alwaysTemplate)
+        view.contentMode = .scaleAspectFit
+    }
+    private lazy var progressContainerView: UIView = .build { view in
+        view.layer.cornerRadius = UX.progressViewAnimatedBackgroundSize / 2.0 * self.computeShrinkingFactor()
+        view.backgroundColor = .clear
+    }
+    private weak var scrollView: UIScrollView?
+    private var obeserveTicket: NSKeyValueObservation?
+    private var currentTheme: Theme?
+    private var refreshIconHasFocus = false
+    private lazy var easterEggGif: UIImageView? = {
+        let imageView = loadGifFromBundle(named: "easterEggGif")
+        imageView?.isHidden = true
+        imageView?.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
+    private var isIpad: Bool {
+        traitCollection.userInterfaceIdiom == .pad && traitCollection.horizontalSizeClass == .regular
+    }
+
+    init(parentScrollView: UIScrollView?,
+         isPotraitOrientation: Bool,
+         onRefreshCallback: @escaping VoidReturnCallback) {
+        self.scrollView = parentScrollView
+        self.onRefreshCallback = onRefreshCallback
+        super.init(frame: .zero)
+        clipsToBounds = true
+        setupEasterEgg(isPotrait: isPotraitOrientation)
+        setupSubviews()
+        startObservingContentScroll()
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    private func setupEasterEgg(isPotrait: Bool) {
+        if let easterEggGif {
```

Comment:
A `guard let` is more suitable here since we can't do any of this work without the gif.